### ISSUE TITLE
Nested Routing

### ISSUE THREAD

**Issue by ScriptedAlchemy** [respond-framework/rudy#3]
Like react router, the ability to have sub routers with their own routes

**Comment by hedgepigdaniel** [comment]
Things to investigate in the codebase:

- createScene (dynamically creates a sort of sub router) - See some WIP code using it: https://github.com/respond-framework/rudy/compare/nested-routes?expand=1
- Is is possible to concatenate regexes with `rudy-match-path` or whatever is being used to do the URL matching?

**Comment by hedgepigdaniel** [comment]
The way I see it working:

### routesMap => baseRoute + children

Instead of the existing flat routesMap and special global callbacks
```javascript
connectRoutes({
  // routeMap (flat)
  ROUTE1: {
    onBeforeChange,
    onAfterChange,
  },
  // ... other routes
}, {
  // Global callbacks
  onBeforeChange,
  onAfterChange,
  // ... other options
});
```

We can introduce the concept of `route.children` and in doing so dispense with the special handling of global callbacks:

```javascript
connectRoutes({
  // Lets call it "baseRoute" - the single root route of the app
  path: '/', // I guess this is basename
  onBeforeChange: loadGlobalData, // This is like the global onBeforeChange option
  onAfterChange: analyticsPageEnter, // This is like the global onAfterChange option
  idKey: 'type', // The "scene name" or the key inside actions that the "id" field goes into (default: "type")
  children: [
    // This is like the original routesMap - the children of the base route
    {
      id: 'POST',
      priority: 2, // This resolves the issue of ambiguous route precedence when paths overlap (hence why children is an array, unlike the original routesMap
      path: 'posts/:postId', // This matches after the parent route = e.g. '/' + 'posts/:id'
      onBeforeChange, // Like the per-route onBeforeChange
      onAfterChange, // Like the per-route onAfterChange
    },
    {
      id: 'ADMIN',
      path: 'admin/',
      onBeforeChange: loadAdminData,
      idKey: 'adminSection',
      children: [ // children can be nested as deep as you want
        {
          id: 'USER',
          path: 'users/:userId', // Matches '/' + 'admin/' + 'users/:userId'
          onBeforeChange: loadAdminUsersData,
        },
        {
          id: 'SUPER',
          path: '/super', // Matches just '/super' (because it starts with '/')
          onBeforeChange: loadSuperUserUiData,
        },
      ],
    },
  ],
}, {
  // ... other options
});
```

So if you navigated to `/admin/users/7`, the action would look something like this:
```javascript
{
  type: 'ADMIN', // Its up to the user what to do with "idKey" to decide which nested id they want to be the action type
  adminSection: 'USER', // All the other scene names are flat and easy to switch on in reducers
  payload: {
    userId: '7', // This would include the payload params from all nested routes
  },
}
```

Before entering the route, `loadGlobalData`, `loadAdminData`, and `loadAdminUsersData` would be executed in parallel. After entering, `analyticsPageEnter` would run.

Alternatively for `/super`, it would look like this:
```javascript
{
  type: 'ADMIN',
  adminSection: 'SUPER',
}
```

### Importing 3rd party sub-apps

With child routes and the concept of nested controllers, maybe its possible to factor out a whole tree of routes in an app into a library? You could publish a package that provides a blog app, and mount it at a path in a bigger app:

```javascript
import { createBlog } from 'blog-library';

const {
  reducer: blogReducer,
  routes: blogRoutes,
  Component: BlogComponent,
} = createBlog({
  reducerName: 'blog',
  idPrefix: 'blog',
});

const {
  reducer: rudyReducer,
  middleware: rudyMiddleware,
} = connectRoutes({
  children: [
    {
      id: 'BLOG',
      children: blogRoutes,
    }
  ],
});

// Create the store, including the blogReducer
// Render the react tree, including mounting <BlogComponent /> on the 'BLOG' route
// The rest is handled by blog-library
```

The actions used by the blog would be something like:
```javascript
{
  type: 'BLOG',
  'blog/type': 'BLOG_LIBRARY_POST', // Whatever the library users for its idKey/id, but with the key prefixed by blog
}
```

I'm a bit stuck on how to avoid the blog-library needing to know about all the parent routes (and therefore what other scene keys and payload params need to go into the action. Maybe actions dispatched can be intercepted and filled in with data from parent routes that the library shouldn't need to know about?

### Other thoughts

- Perhaps the route callbacks should receive a `selfChanged` parameter in the `bag` to indicate whether the match for that route has changed, or only for one of the child routes. That would simplify the option of only calling thunks in routes if the part of he URL matched by that route has changed.
- By putting the `idKey` in the route options, its possible to have different keys/names for the id's of different sets of child routes. For example, you could have a route `MARKETING` as a child of the base route, and its children could use `page` as the key (not necessarily `adminSection`).
- To generate valid redux actions, one of the keys in all actions must be `type` Perhaps we should enforce that the base route's idKey is `type`, or otherwise have an automated mechanism to ensure that there is an implicit `NOT_FOUND` route at all levels.
- We don't need separate code to handle global route callbacks and nested route callbacks this way
- I haven't thought about it too much but I think you could probably neatly code split child routes. Maybe all you need is `ADD_ROUTES` to work with children?
- Not sure how simple it is to concatenate the paths. What if there are parent and child routes like `/:pathPath+` and `/:subPagePath+`. How to decide for the URL `/one/two/three` whether its `/one/two` and `/three` or `/one` and `/two/three`? Just concatenate the regexes and leave it to the user?

**Comment by faceyspacey** [comment]
Quick thing: people love types as keys instead of arrays. There is no priority problem. All browsers respect the order of keys in objects, as long as numbers aren’t used.

Priority is also more work. The goal is always more power with less.

**Comment by faceyspacey** [comment]
Concatenating paths may require us to alert the user at runtime (during DEV only) if parent routes cannot be appended to. That’s a fine solution in my book.

**Comment by faceyspacey** [comment]
As far as idKey, basically scenes addresses the same goal. Here’s why: in our system in a sufficiently complex/nested app, u often will match on scene, rather than type. So you will create something very much like a module by using createScene. All its routes will have the same scene.

We will have to address how scenes from children associate with a parent scene. Which is shown? The goal is a representation of all tiers of scenes.

But this is part 2. Don’t worry about it yet. The perfect solution will come just by meditating on what I just said while diving deeper via part 1.

**Comment by faceyspacey** [comment]
Basically remixx has 2 strategies in mind for imported arbitrary 3rd party modules: 

- appending guaranteed to be unique hash strings to action types (and state keys)
- or letting users assign them (as described in the last comment)

There is some problems with both approaches, but when I/we finalize the chosen strategy, it will solve the long time Achilles heel of redux: that it’s not composable. 

See, what we are doing is cosmic; the React team has invested tens of millions of dollars of mark zuckerberg’s money on Facebook employee salaries to make the lowest level primitive, components, be able to do all things. Hooks and suspense are no accident. Their purpose is for React to be all u need. And to achieve the goal of App = f(state) like us. But not quite. 

After React 17 comes out, the empire will strike back (us lol) with a higher level abstraction/primitive: remixx modules. Basically they bundle components, reducers, routes (and therefore actions) into one grouping which can be hosted on npm as a 3rd party module, and then imported into ur app collision-free.

So bottom line: our redux killer will have a close coupling to Rudy, and the collision strategy invented then will be applied back to here. The end.

**Comment by hedgepigdaniel** [comment]
Thanks, that's alot of good stuff to digest, but I feel like we're pretty synced on the important short term things. The idea of implementing the callback part of nested routes first is a good one I think - that part I think might actually simplify the code, and then handling nested paths would be a logical next step.

**Comment by hedgepigdaniel** [comment]
Re the route ordering issue - I know objects are nice, but I'm doubtful about relying on undefined behaviour that just happens to be the same across browsers, has surprising special behaviour for numbers, isn't documented, etc. I prefer the idea that works in a simple predictable way, and guaranteed to work in any spec compliant browser, even if that means the code is a little more verbose.

Another issue is how to handle the case where routes are added dynamically. What if multiple sets of overlapping routes are added dynamically (maybe in different orders)?

I definitely agree that doing extra work to implement priority is unnecessary at this point (IMO, a well designed routesMap does not have overlapping routes in the first place). I'm just thinking that since nested routes is a significant refactor of the shape of routesMap, we may aswell switch to arrays now and put the ambiguous/unpredictable ordering issue to bed, even if we don't implement priority yet.

**Comment by hedgepigdaniel** [comment]
> If changes are necessitated here, we aren’t afraid of major version bumps. Our mantra by the way is perfecting the vision, not perfect backward compatibility like React itself which progresses at a snails pace. We don’t care about being the most popular. That’s a by product.

Absolutely agree. I don't think we need to follow the path of absolutely minimal breaking changes - not least because we don't have a facebook sized infrastructure lol.

I think it would be a healthy thing to do some releases along the way though. The current codebase solves a huge number of problems that RFR doesn't (async call middleware, hash handling, middleware API, etc).

The sooner we get that out to people, the more help we get with bugfixes and the like, and more more people move towards our way of thinking, start building an ecosystem of middlewares, etc. We shouldn't focus on popularity as an end in itself, and we definitely shouldn't slow down development to avoid breaking changes. That said, I think releasing incrementally as we go is a good thing. There's really nothing to lose and much to gain. If nothing else, we'll hopefully get a bit of encouragement from curious alpha testers, and thats a good thing.

**Comment by hedgepigdaniel** [comment]
Haskell is beautiful. Its amazing how much pain people go through in JS to get to the most basic/standard benefits of almost anything written in Haskell.

I've spent so many hours of my life configuring webpack in various projects so that its a compile error if you import a name that is completely undefined. Its good to have it working but wow, I can't believe that still today is usually not the case.

**Comment by ScriptedAlchemy** [comment]
Regarding releases. I’ll agree with a silent - / private release for our sakes. 

Regarding typescript, I think holding the project back is not worth just changeijg from flow to TS, that can be done under the hood and we will have community support in writing the definitions. 

I’d like to consider riding the hype wave that’s being built. There’s high visibility and the community kinda knows we are one major feature away. 

Deferring it further because we wanna switch to TS risks us losing the traction. 

Moving to TS is really something we can do under the hood. 

Rudy is good enough that people will volenteer and have. RFR2 docs were community driven. The most important thing is nested routes and push a beta or something privately so we can test and others we select can trial. Specifically some enterprise companies who might be willing to risk the upgrade. 

I don’t want limited resources to be strained over semantics. There’s lower level developers who’d be willing to help with something easy, like moving to TS. 

I don’t know TS but will learn it. However - it’s time We could spend documentation and marketing. 

Think about it bro :) we aren’t gonna release and stop maintaining. Flow is kinda frustrating so I’m down to move. Just not when we are on the cusp of being ready, it’s been hard to restore faith that it’s not vaporware, getting ready for release and dropping it’s last killer feature should be high priority. The community is updating RFR TS type defs (apparently they exist somewhere). 

Something to consider - I intend rolling Rudy to angular and vue at some point. TS will be good. But we shouldn’t wait for the next wave especially with React 17. I feel we could loose steam if we delay this till React is just async. It’s less boom then we get now.

**Comment by ScriptedAlchemy** [comment]
Universal is already taking a knock. Timing is everything. I don’t want us to be in limbo - irrelevant for a period of time. 

We almost lost Universal last year, we clawed back from 3k download a month to 20k

**Comment by faceyspacey** [comment]
Master this entire codebase, then we talk. Nothing is gonna implode, we don’t gotta freak out. Fine on the TS.

**Comment by hedgepigdaniel** [comment]
Ok, I kind of see where you coming from re the object routesMap. I think its actually possible to fix the ambiguous ordering without changing to arrays. Basically (at some point, no a priority), we can internally define the order to be something stable (say alphabetical order) and introduce a priority field so that users can specify priorities if need be. That way we can keep the nice looking object/docs, and for apps that for some reason need to have overlapping routes its possible to make it work in a predictable way. So I think we can leave i the way it is for the time being.

Re releases and community expectations, alot has to do with messaging. At the end of the day, this is all in public and MIT licenced, so anyone is free to checkout out the code and build it, and release it under a different name if they so choose. I think the more important issue is how we market it. I agree that releasing 1.0 final and saying "its ready" and advertising it everywhere is foolish until its right, at least by our standards, and we all understand what it is we are claiming is ready. But if we release something early, call it an alpha, make it clear that its incomplete, probably buggy, probably gonna have breaking changes tomorrow, then that should temper people's expectations. At the same time, we will probably get more community involvement and help - and that will help us get to the 1.0 goal.

So I say we put it on NPM as a pre 1.0 version with say an alpha tag, and/or do a release for each push to master tagged with a commit hash (lerna + travis should make that pretty easy). We don't need to announce that its "released" or "ready" or "done" or announce in the README that its on npm, just make it simple to try out for those that are searching deep and want to help/experiment. That includes us - I'd like to be able to easily try it on various apps that I work on without getting lost on dodgy git/building problems or NPM access/permissions.

**Comment by hedgepigdaniel** [comment]
I'll try and make a start on nested routes soon. Its good that you're here James. If you're able to review it as we go then that will make me alot more confident in making that sort of significant change in the right way. I'll open a PR once I have something that makes sense.

**Comment by ScriptedAlchemy** [comment]
Agreed, very good to see the man himself back on a machine. Ur feedback here was critical

**Comment by hedgepigdaniel** [comment]
Ok I've started work on this, WIP on this branch: https://github.com/respond-framework/rudy/compare/nested-routes?expand=1

I'm planning to follow the implementation plan from James's [comment](https://github.com/respond-framework/rudy/issues/3#issuecomment-436517049) - so the first step is for all child routes to have absolute paths, and mostly worry about the nested callbacks.
